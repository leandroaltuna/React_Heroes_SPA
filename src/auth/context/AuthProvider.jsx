import { useReducer } from 'react';
import { AuthContext } from './AuthContext';
import { authReducer } from './authReducer';
import { types } from '../types';


//* El init es el tercer param donde se puede inicializar el state.
const init = () => {
    
    const userStoraged = JSON.parse( localStorage.getItem( 'userStoraged' ) );

    return {
        logged: !!userStoraged, //? Doble negacion para saber si hay info del usuario almacenado en el localstorage.
        userInfo: userStoraged,
    }

}

// const initialState = {
//     logged: false,
// }

export const AuthProvider = ({ children }) => {
  
    //* El init es el tercer param del useReducer hook que puede ser opcional. En este caso lo usamos para poder guardar los datos del usuario en el localstorage.
    const [ authState, dispatch ] = useReducer( authReducer, {}, init );
    // const [ authState, dispatch ] = useReducer( authReducer, initialState );

    const login = ( name = '' ) => {
        
        const userObj = { id: 'ABC', name: name, };

        const action = {
            type: types.login,
            payload: userObj,
            // payload: {
            //     id: 'ABC',
            //     name: name,
            // }
        };

        localStorage.setItem( 'userStoraged', JSON.stringify( userObj ) );

        dispatch( action );

    }

    const logout = () => {
        
        localStorage.removeItem( 'userStoraged' );

        const action = { type: types.logout, };

        dispatch( action );

    }

    return (
        <AuthContext.Provider 
            value={
                {
                    ...authState,
                    login,
                    logout,
                }
            }
        >
            { children }
        </AuthContext.Provider>
    )

}
