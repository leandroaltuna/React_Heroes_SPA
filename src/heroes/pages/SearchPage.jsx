import { useLocation, useNavigate } from 'react-router-dom';
import queryString from 'query-string';
import { useForm } from '../../hooks';
import { getHeroesByName } from '../helpers';
import { HeroItem } from '../components/HeroItem';


export const SearchPage = () => {
    
    //* Obtener la navegacion
    const navigate = useNavigate();
    //* Obtener informacion de donde nos encontramos
    const location = useLocation();

    //* Cambio el nombre de q a queryParam y lo asigno '' por defecto. query-string es una libreria para obtener los parametros que se envian por url.
    const { q : queryParam = '' } = queryString.parse( location.search );
    const heroes = getHeroesByName( queryParam );

    //* Las condiciones ya regresan un valor booleano, por ende no es necesario asignar el valor booleano en el codigo.
    const showSearchAlert = ( queryParam.length === 0 );
    const showErrorAlert = ( queryParam.length > 0 ) && heroes.length === 0;


    const { searchText, onInputChange } = useForm({
        // searchText: '',
        searchText: queryParam,
    });
    

    const onSearchSubmit = ( event ) => {
        
        event.preventDefault();

        // if ( searchText.trim().length <= 1 ) return;

        navigate( `?q=${ searchText }` );

    }
    
    
    return (
        <>
            <h1>Search</h1>
            <hr />

            <div className="row">

                <div className="col-5">
                    
                    <h4>Searching</h4>
                    <hr />

                    <form onSubmit={ onSearchSubmit }>
                        <input
                            className="form-control" 
                            type="text" 
                            name="searchText" 
                            placeholder="Search a hero"
                            autoComplete="off"
                            value={ searchText }
                            onChange={ onInputChange }
                        />

                        <button className="btn btn-outline-primary mt-2">
                            Search
                        </button>
                    </form>

                </div>

                <div className="col-7">
                    
                    <h4>Results</h4>
                    <hr />

                    
                    {/* //* Si no busca nada se mostrara el search alert, caso contrario se oculta. */}
                    <div 
                        className="alert alert-primary animate__animated animate__fadeIn" 
                        style={{ display: (showSearchAlert) ? '' : 'none' }}
                    >
                        Search a hero
                    </div>
                   
                    {/* //* Si la busqueda no retorna ningun valor se mostrara el error alert, caso contrario se oculta. */}
                    <div 
                        className="alert alert-danger animate__animated animate__fadeIn" 
                        style={{ display: (showErrorAlert) ? '' : 'none' }}
                    >
                        No Hero with <b>{ queryParam }</b>
                    </div>

                    {
                        heroes.map( ( hero ) => (
                            <HeroItem 
                                key={ hero.id }
                                { ...hero }
                            />
                        ))
                    }

                </div>

            </div>
        </>
    )
}
