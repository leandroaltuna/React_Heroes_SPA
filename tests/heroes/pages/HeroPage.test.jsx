import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter, Route, Routes, useNavigate, useParams } from 'react-router-dom';
import { HeroPage } from '../../../src/heroes/pages/HeroPage';
import { AuthContext } from '../../../src/auth/context';


const mockedUseNavigate = jest.fn();
const mockedUseParams = jest.fn();

const heroId = 'dc-flash';


jest.mock( 'react-router-dom', () => ({

    ...jest.requireActual( 'react-router-dom' ),
    useNavigate: () => mockedUseNavigate,
    useParams: () => mockedUseParams,
    // useParams: () => ({ heroId }),

}));


describe( 'Pruebas en <HeroPage />', () => { 
    
    beforeEach( () => jest.clearAllMocks() );

    test( 'Debe mostrar info del Hero obtenido por url', () => { 
    
        jest.spyOn( require('react-router-dom'), 'useParams').mockReturnValue({ heroId });

        const { container } = render(
            <MemoryRouter>
                <HeroPage />
            </MemoryRouter>
        );

        // screen.debug();
        expect( container ).toMatchSnapshot();
        
    });

    test( 'Debe mostrar la imagen del Hero obtenido por url', () => { 
        
        jest.spyOn( require('react-router-dom'), 'useParams').mockReturnValue({ heroId });

        render(
            <MemoryRouter>
                <HeroPage />
            </MemoryRouter>
        );

        const heroImg = screen.getByRole( 'img' );

        expect( heroImg.src ).toContain( `/heroes/${ heroId }.jpg` );

    });

    test( 'Debe de navegar a la pagina anterior cuando hace click en boton Back', () => { 

        jest.spyOn( require('react-router-dom'), 'useParams').mockReturnValue({ heroId });

        render(
            <MemoryRouter>
                <HeroPage />
            </MemoryRouter>
        );

        // screen.debug();
        const backBtn = screen.getByRole( 'button' );
        fireEvent.click( backBtn );

        expect( mockedUseNavigate ).toHaveBeenCalledWith( -1 );
        
    });

    test( 'Debe de navegar al /marvel si no es el heroId correcto', () => { 
        

        jest.spyOn( require('react-router-dom'), 'useParams').mockReturnValue({ heroId: 'marvel-captainxx' });

        render(

            <AuthContext.Provider value={{ logged: true }}>
                <MemoryRouter initialEntries={[ '/hero' ]}>
                    <Routes>

                        <Route path='hero' element={ <HeroPage /> } />
                        
                        <Route path='marvel' element={
                            <h1>Marvel Page</h1>
                        } />

                    </Routes>
                </MemoryRouter>
            </AuthContext.Provider>

        );

        expect( screen.getByText( 'Marvel Page' ) ).toBeTruthy(); 
        
    });
    
});