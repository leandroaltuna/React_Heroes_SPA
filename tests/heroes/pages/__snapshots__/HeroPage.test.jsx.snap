// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <HeroPage /> Debe mostrar info del Hero obtenido por url 1`] = `
<div>
  <div
    class="row mt-5"
  >
    <div
      class="col-4 animate__animated animate__fadeInLeft"
    >
      <img
        alt="Flash"
        class="img-thumbnail"
        src="/heroes/dc-flash.jpg"
      />
    </div>
    <div
      class="col-8"
    >
      <h3>
        Flash
      </h3>
      <ul
        class="list-group list-group-flash"
      >
        <li
          class="list-group-item"
        >
          <b>
            Alter Ego:
          </b>
           
          Jay Garrick
        </li>
        <li
          class="list-group-item"
        >
          <b>
            Publisher:
          </b>
           
          DC Comics
        </li>
        <li
          class="list-group-item"
        >
          <b>
            First Appearance:
          </b>
           
          Flash Comics #1
        </li>
      </ul>
      <h5
        class="mt-3"
      >
        Characters:
      </h5>
      <p>
        Jay Garrick, Barry Allen, Wally West, Bart Allen
      </p>
      <button
        class="btn btn-outline-primary"
      >
        Back
      </button>
    </div>
  </div>
</div>
`;
